simple/utils/spectral_types.py

```diff
@@ -3,320 +3,317 @@
+        "comments": comments,
+        "reference": reference,
+    }
+
+    logger.debug(f"Trying to insert {spt_data} into Spectral Types table ")
+
+    try:
+        spt_obj = SpectralTypes(**spt_data)
+        with db.session as session:
+            session.add(spt_obj)
+            session.commit()
+        logger.info(f"Spectral type added to database: {spt_data}\n")
+
+        # unset old adopted only after ingest is successful!
+        if adopted and old_adopted is not None:
             with db.engine.connect() as conn:
-                conn.execute(db.SpectralTypes.insert().values(spt_data))
+                conn.execute(
+                    db.SpectralTypes.update()
+                    .where(
+                        and_(
+                            db.SpectralTypes.c.source == old_adopted["source"][0],
+                            db.SpectralTypes.c.regime == old_adopted["regime"][0],
+                            db.SpectralTypes.c.reference == old_adopted["reference"][0],
+                        )
+                    )
+                    .values(adopted=False)
+                )
                 conn.commit()
-            n_added += 1
-            msg = f"Added {str(spt_data)}"
-            logger.debug(msg)
-        except sqlalchemy.exc.IntegrityError as e:
-            if (
-                db.query(db.Publications)
-                .filter(db.Publications.c.reference == references[i])
-                .count()
-                == 0
-            ):
-                msg = f"The publication does not exist in the database: {references[i]}"
-                msg1 = "Add it with ingest_publication function."
-                logger.debug(msg + msg1)
-                raise AstroDBError(msg)
-            elif "NOT NULL constraint failed: SpectralTypes.regime" in str(e):
-                msg = f"The regime was not provided for {source}"
+    except sqlalchemy.exc.IntegrityError as e:
+        if (
+            db.query(db.Publications)
+            .filter(db.Publications.c.reference == reference)
+            .count()
+            == 0
+        ):
+            msg = f"The publication does not exist in the database: {reference}"
+            msg1 = "Add it with astrodb_utils.ingest_publication function."
+            logger.debug(msg + msg1)
+            if raise_error:
                 logger.error(msg)
-                raise AstroDBError(msg)
+                raise AstroDBError(msg + msg1)
             else:
-                msg = "Other error\n"
+                logger.warning(msg + msg1)
+        else:
+            msg = f"Spectral type ingest failed with error {e}\n"
+            if raise_error:
                 logger.error(msg)
                 raise AstroDBError(msg)
+            else:
+                logger.warning(msg)
 
-    msg = f"Spectral types added: {n_added} \n" f"Spectral Types skipped: {n_skipped}"
-    logger.info(msg)
+    # check that adopted flag is successfully changed
```

Comment:
Should these checks/validation just be part of the tests? In fact, I'm pretty sure we already have tests for this. That would make this function simpler to parse and unit test.

Reply:
If you want to have it in spectral_type.py, they should be its own separate function (which can be called by `ingest_spectral_types` if you really need to).